File: Python/16. 3SumClosest.py
'''
Given an array nums of n integers and an integer target, find three integers in nums such that the sum is closest to target. Return the sum of the three integers. You may assume that each input would have exactly one solution.

Example:
Input: nums = [-1,2,1,-4], target = 1
Output: 2
Explanation: The sum that is closest to the target is 2. (-1 + 2 + 1 = 2).
'''

class Solution(object):
    def threeSumClosest(self, nums, target):
        m = 20000
        l = len(nums)
        for i in range(0, l-2):
            for j in range(i+1, l-1):
                for k in range(j+1, l):
                    s = nums[i] + nums[j] + nums[k]
                    diff = abs(s-target)
                    if m>diff:
                        m = diff
                        closest_sum = s
        return(closest_sum)

File: C++/16. 3SumClosest.cpp
//Problem 16
// Find the closest sum of 3 numbers out of all the elements of an array given a target input





class Solution {
public:
    int threeSumClosest(vector<int>& nums, int target) {
        int n = nums.size();
        int i = 0;
        int closestSum = nums[0] + nums [1] + nums[n-1];   //Initially closestSum = initial triplet sum
        sort(nums.begin(),nums.end());
        while(i < n -2){
            int j = i+1;
            int k = n-1;         
            while(j < k){
                int sum = nums[i] + nums[j] + nums[k] ;
                if(sum == target)               //if triplet sum becomes equal to target, then that is the closestSum itself
                    return sum;
                if (abs(target - sum) < abs(target - closestSum)){
                    closestSum = sum;       //if sum of current triplet is closer than target than the previous closestSum,
				}                                        //we update the closestSum value to the current triplet sum
                if(sum > target)
                    k--;
                else if(sum <target)
                    j++;
            }
            i++;
        }
        return closestSum;
    }
};
